src/passes/Heap2Local.cpp

```diff
@@ -0,0 +1,617 @@
+            builder.makeConstantExpression(Literal::makeZero(fields[i].type))));
+        }
+      }
+
+      // Put the allocation itself at the end of the block, so the block has the
+      // exact same type as the allocation it replaces.
+      contents.push_back(allocation);
+      replaceCurrent(builder.makeBlock(contents));
+    }
+
+    void visitStructSet(StructSet* curr) {
+      if (!reached.count(curr)) {
+        return;
+      }
+
+      // Drop the ref (leaving it to other opts to remove, when possible), and
+      // write the data to the local instead of the heap allocation.
+      replaceCurrent(builder.makeSequence(
+        builder.makeDrop(curr->ref),
+        builder.makeLocalSet(localIndexes[curr->index], curr->value)));
+    }
+
+    void visitStructGet(StructGet* curr) {
+      if (!reached.count(curr)) {
+        return;
+      }
+
+      replaceCurrent(
+        builder.makeSequence(builder.makeDrop(curr->ref),
+                             builder.makeLocalGet(localIndexes[curr->index],
+                                                  fields[curr->index].type)));
+    }
+  };
+
+  enum class ParentChildInteraction {
+    // The parent lets the child escape. E.g. the parent is a call.
+    Escapes,
+    // The parent fully consumes the child in a safe, non-escaping way, and
+    // after consuming it nothing remains to flow further through the parent.
+    // E.g. the parent is a struct.get, which reads from the allocated heap
+    // value and does nothing more with the reference.
+    FullyConsumes,
+    // The parent flows the child out, that is, the child is the single value
+    // that can flow out from the parent. E.g. the parent is a block with no
+    // branches and the child is the final value that is returned.
+    Flows,
+    // The parent does not consume the child completely, so the child's value
+    // can be used through it. However the child does not flow cleanly through.
+    // E.g. the parent is a block with branches, and the value on them may be
+    // returned from the block and not only the child. This means the allocation
+    // is not used in an exclusive way, and we cannot optimize it.
+    Mixes,
+  };
+
+  // Analyze an allocation to see if we can convert it from a heap allocation to
+  // locals.
+  bool convertToLocals(StructNew* allocation) {
+    Rewriter rewriter(allocation, func, module);
+
+    // A queue of flows from children to parents. When something is in the queue
+    // here then it assumed that it is ok for the allocation to be at the child
+    // (that is, we have already checked the child before placing it in the
+    // queue), and we need to check if it is ok to be at the parent, and to flow
+    // from the child to the parent. We will analyze that (see
+    // ParentChildInteraction, above) and continue accordingly.
+    using ChildAndParent = std::pair<Expression*, Expression*>;
+    UniqueNonrepeatingDeferredQueue<ChildAndParent> flows;
```

Comment:
Oh, that is actually not crucial. The nonrepeating part is essential, but the order should not matter here.

Parents must always be visited after children by design in this flow, since it is from children to parents.

There can be cycles in theory since we follow the paths of locals and branches, and those are checked for exclusivity. GIven that property, it does not matter in what order we visit them.

(The utility class defers rather than does things more eagerly because in other places that was benchmarked as faster, that's it really.)